rules/abstract-rector/src/Rector/AbstractConvertToAnonymousFunctionRector.php

```diff
@@ -0,0 +1,108 @@
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\Assign;
+use PhpParser\Node\Expr\Closure;
+use PhpParser\Node\Expr\ClosureUse;
+use PhpParser\Node\Expr\Variable;
+use PhpParser\Node\Param;
+use PhpParser\Node\Stmt;
+use PhpParser\Node\Stmt\Expression;
+use Rector\Core\Rector\AbstractRector;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+
+/**
+ * @see https://www.php.net/functions.anonymous
+ */
+abstract class AbstractConvertToAnonymousFunctionRector extends AbstractRector
+{
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkip($node)) {
+            return null;
+        }
+
+        $parameters = $this->getParameters($node);
+        $body = $this->getBody($node);
+        $useVariables = $this->resolveUseVariables($body, $parameters);
+
+        $anonymousFunctionNode = new Closure();
+
+        foreach ($parameters as $parameter) {
+            $anonymousFunctionNode->params[] = new Param($parameter);
+        }
+
+        if ($body !== []) {
+            $anonymousFunctionNode->stmts = $body;
+        }
+
+        foreach ($useVariables as $useVariable) {
+            $anonymousFunctionNode->uses[] = new ClosureUse($useVariable);
+        }
+
+        return $anonymousFunctionNode;
+    }
+
+    abstract protected function shouldSkip(Node $node): bool;
+
+    /**
+     * @return Variable[]
+     */
+    abstract protected function getParameters(Node $node): array;
+
+    /**
+     * @return Expression[]|Stmt[]
+     */
+    abstract protected function getBody(Node $node): array;
+
+    /**
+     * @param Node[] $nodes
+     * @param Variable[] $paramNodes
+     * @return Variable[]
+     */
+    private function resolveUseVariables(array $nodes, array $paramNodes): array
+    {
+        $paramNames = [];
+        foreach ($paramNodes as $paramNode) {
+            $paramNames[] = $this->getName($paramNode);
+        }
+
+        $variableNodes = $this->betterNodeFinder->findInstanceOf($nodes, Variable::class);
+
+        /** @var Variable[] $filteredVariables */
+        $filteredVariables = [];
+        $alreadyAssignedVariables = [];
+        foreach ($variableNodes as $variableNode) {
+            // "$this" is allowed
+            if ($this->isName($variableNode, 'this')) {
+                continue;
+            }
+
+            $variableName = $this->getName($variableNode);
+            if ($variableName === null) {
+                continue;
+            }
+
+            if (in_array($variableName, $paramNames, true)) {
+                continue;
+            }
+
+            $parentNode = $variableNode->getAttribute(AttributeKey::PARENT_NODE);
+            if ($parentNode instanceof Assign) {
+                $alreadyAssignedVariables[] = $variableName;
+            }
+
+            if ($this->isNames($variableNode, $alreadyAssignedVariables)) {
+                continue;
+            }
+
+            $filteredVariables[$variableName] = $variableNode;
+        }
+
```

Comment:
What does this foreach do? Could you extract it to own method, so the name tells us?

Reply:
This foreach is calculating the variables that will go under `use`:
```php
function ($var1) use ($var2) {
```

For instance, if when parsing the function it finds variable `$foo`, then it checks that it's not a parameter to the function, and that it was not first assigned in the function, then the variable must've been defined outside and must go to `use`.

So I think there's no need to extract it, since it's already doing `resolveUseVariables`